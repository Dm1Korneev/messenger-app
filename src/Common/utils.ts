
type Filter = Parameters<typeof Array.prototype.filter>[0]
export const distinct: Filter = (value, index, self) => self.indexOf(value) === index;
