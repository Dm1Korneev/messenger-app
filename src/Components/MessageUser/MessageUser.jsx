import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import ListItem from '@material-ui/core/ListItem';
import Typography from '@material-ui/core/Typography';

import UsersAvatar from 'Components/UsersAvatar';

const useStyles = makeStyles((theme) => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 999,
    padding: theme.spacing(0, 2),
  },
}));

const MessageUser = (props) => {
  const classes = useStyles();

  const {
    isCurrentUserMessage, author, avatar, children,
  } = props;

  const flexDirection = isCurrentUserMessage ? 'row-reverse' : 'row';

  return (
    <ListItem>
      <Box
        display="flex"
        flex="1"
        alignItems="start"
        flexDirection={flexDirection}
      >
        <Box position="sticky" top={0}>
          <UsersAvatar author={author} avatar={avatar} />
        </Box>
        <Box display="flex" flex="1" flexDirection="column">
          <Box
            position="sticky"
            top={0}
            display="flex"
            flexDirection={flexDirection}
            className={classes.ListItemText}
          >
            <Typography
              component="span"
              variant="subtitle2"
            >
              {author}
            </Typography>
          </Box>
          {children}
        </Box>
      </Box>
    </ListItem>
  );
};

MessageUser.propTypes = {
  isCurrentUserMessage: PropTypes.bool.isRequired,
  author: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
};

export default MessageUser;

