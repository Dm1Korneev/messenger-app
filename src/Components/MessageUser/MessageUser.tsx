import Box from '@material-ui/core/Box';
import ListItem from '@material-ui/core/ListItem';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { FC } from 'react';

import { UsersAvatar } from 'Components/UsersAvatar';

const useStyles = makeStyles((theme) => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 999,
  },
}));

type Props = {
  isCurrentUserMessage: boolean;
  author: string;
  avatar: string;
}

export const MessageUser: FC<Props> = ({
  isCurrentUserMessage, author, avatar, children,
}) => {
  const classes = useStyles();

  const flexDirection = isCurrentUserMessage ? 'row-reverse' : 'row';

  return (
    <ListItem>
      <Box
        display="flex"
        flex="1"
        alignItems="start"
        flexDirection={flexDirection}
      >
        <Box position="sticky" top={0}>
          <UsersAvatar author={author} avatar={avatar} />
        </Box>
        <Box display="flex" flex="1" flexDirection="column" px={2}>
          <Box
            position="sticky"
            top={0}
            display="flex"
            flexDirection={flexDirection}
            className={classes.ListItemText}
          >
            <Typography
              component="span"
              variant="subtitle2"
            >
              {author}
            </Typography>
          </Box>
          {children}
        </Box>
      </Box>
    </ListItem>
  );
};
