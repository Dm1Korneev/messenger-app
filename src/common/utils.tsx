export function distinct(value: any, index: number, self: Array<any>): boolean {
  return self.indexOf(value) === index;
}
