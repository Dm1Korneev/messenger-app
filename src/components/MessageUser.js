import React from "react";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import withStyles from "@material-ui/core/styles/withStyles";
import Typography from "@material-ui/core/Typography";
import classNames from "classnames";
import UsersAvatar from "./UsersAvatar";

function MessageUser(props) {
  const { isCurrentUserMessage, author, avatar, children, classes } = props;

  const ListItemTextClass = classNames(
    classes.ListItemText,
    isCurrentUserMessage && classes.reverse
  );

  const ListItemClass = classNames(
    classes.ListItem,
    isCurrentUserMessage && classes.reverse
  );

  return (
    <ListItem className={ListItemClass}>
      <div className={classes.sticky}>
        <UsersAvatar author={author} avatar={avatar} />
      </div>
      <div className={classes.textContent}>
        <ListItemText
          className={classNames(ListItemTextClass, classes.sticky)}
          primary={
            <Typography
              component="span"
              className={classes.ListItemText__author}
            >
              {author}
            </Typography>
          }
        />
        {children}
      </div>
    </ListItem>
  );
}

const styles = theme => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 999,
    display: "flex",
    justifyContent: "space-between",
    "&:first-child": {
      paddingLeft: theme.spacing.unit * 2
    }
  },
  ListItemText_reverse: {
    display: "flex",
    flexDirection: "row-reverse",
    justifyContent: "space-between"
  },
  reverse: {
    flexDirection: "row-reverse"
  },
  ListItem: {
    alignItems: "start",
    paddingTop: theme.spacing.unit,
    paddingBottom: 0,
    paddingLeft: theme.spacing.unit * 2
  },
  textContent: {
    flexGrow: 1
  },
  ListItemText__author: {
    fontSize: theme.typography.pxToRem(12),
    fontWeight: 700
  },
  sticky: {
    position: "sticky",
    top: 0
  }
});

export default withStyles(styles)(MessageUser);
