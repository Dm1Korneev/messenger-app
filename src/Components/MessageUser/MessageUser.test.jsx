import React from 'react';
import { render } from '@testing-library/react';

import { mockComponent } from 'testing/utils';

import MessageUser from './MessageUser';

jest.mock('Components/UsersAvatar', () => mockComponent('UsersAvatar'));

jest.mock('@material-ui/core/ListItem', () => mockComponent('ListItem'));
jest.mock('@material-ui/core/ListItemText', () => mockComponent('ListItemText'));
jest.mock('@material-ui/core/Typography', () => mockComponent('Typography'));

const scrollIntoViewMock = jest.fn();
window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

const props = {
  isCurrentUserMessage: false,
  author: '',
  avatar: '',
  children: <></>,
};

describe('render MessageUser component', () => {
  const setup = () => render(
    <MessageUser {...props} />,
  );

  test('snapshot test', async () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});
