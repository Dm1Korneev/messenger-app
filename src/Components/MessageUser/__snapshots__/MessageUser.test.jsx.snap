// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MessageUser component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="ListItem"
    test-prop-classname="\\"MessageUser-ListItem-4\\""
  >
    <div
      class="MessageUser-sticky-7"
    >
      <div
        data-originalcomponent="UsersAvatar"
        test-prop-author="\\"\\""
        test-prop-avatar="\\"\\""
      />
    </div>
    <div
      class="MessageUser-textContent-5"
    >
      <div
        data-originalcomponent="ListItemText"
        test-prop-classname="\\"MessageUser-ListItemText-1 MessageUser-sticky-7\\""
        test-prop-primary="<div data-originalcomponent=\\"Typography\\" test-prop-component=\\"&quot;span&quot;\\" test-prop-classname=\\"&quot;MessageUser-ListItemText__author-6&quot;\\" data-reactroot=\\"\\"></div>"
      />
    </div>
  </div>
</DocumentFragment>
`;
